File: src/Calculator/PerfUtils.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Storage;

namespace CalculatorApp
{
    public class PerfUtils
    {
        public class ScopedLog : IDisposable
        {
            public ScopedLog(PerfUtils utils, string message)
            {
                _utils = utils;
                _msg = message;
                _utils.WriteLine($"ScopedLog | {_msg} | begins ");
            }

            ~ScopedLog()
            {
                _utils.WriteLine($"ScopedLog | {_msg} | ends with bad timing");
            }

            public void Dispose()
            {
                _utils.WriteLine($"ScopedLog | {_msg} | ends");
                GC.SuppressFinalize(this);
            }

            private PerfUtils _utils;
            private string _msg;
        }

        public readonly static PerfUtils Default = new PerfUtils();

        public PerfUtils()
        {
            var date = DateTime.UtcNow;
            var filename = $"calc.perflog-{date.ToString("dd-MM-yy_hh-mm-ss-ff")}.log";

            StorageFolder storageFolder =
                ApplicationData.Current.LocalFolder;

            _file = storageFolder.CreateFileAsync(filename, CreationCollisionOption.ReplaceExisting).AsTask().ConfigureAwait(false).GetAwaiter().GetResult();

            _workthread = new Thread(new ThreadStart(_WorkThread));
            _workthread.Start();
        }

        public ScopedLog CreateScopedLog(string content)
        {
            return new ScopedLog(this, content);
        }

        public void WriteLine(string content)
        {
            var date = DateTime.UtcNow;
            var prefix = $"{date.ToString("dd-MM-yy hh:mm:ss.ff")} | ";
            _SubmitReq(new WriteRequest { Content = $"{prefix}{content}\n" });
            lock(_workthread)
            {
                Monitor.Pulse(_workthread);
            }
        }

        public void Quit(bool join = true)
        {
            _quit = true;
            lock(_workthread)
            {
                Monitor.Pulse(_workthread);
            }
            _workthread.Join();
        }

        private struct WriteRequest
        {
            public string Content;
        }

        private void _WorkThread()
        {
            while(!_quit)
            {
                lock (_workthread)
                {
                    while (_TryFetchReq(out WriteRequest? req))
                    {
                        FileIO.AppendTextAsync(_file, req.Value.Content).AsTask().ConfigureAwait(false).GetAwaiter().GetResult();
                    }
                    Monitor.Wait(_workthread, 2000);
                }
            }
        }

        private void _SubmitReq(WriteRequest req)
        {
            lock(_wreqs)
            {
                _wreqs.Enqueue(req);
            }
        }

        private bool _TryFetchReq(out WriteRequest? req)
        {
            lock(_wreqs)
            {
                if(_wreqs.Count > 0)
                {
                    req = _wreqs.Dequeue();
                    return true;
                }
            }

            req = null;
            return false;
        }

        private Windows.Storage.StorageFile _file;

        private Queue<WriteRequest> _wreqs = new Queue<WriteRequest>();
        private Thread _workthread = null;
        private volatile bool _quit = false;
    }
}

File: src/Calculator/PerfUtils.cpp
#include <chrono>

#include "pch.h"
#include "PerfUtils.h"

#include "CalcViewModel/Common/Utils.h"

using namespace Concurrency;
using namespace Windows::Storage;
using namespace Windows::Foundation;

namespace CalculatorApp
{
    PerfUtils PerfUtils::_default;

    PerfUtils& PerfUtils::Default()
    {
        return _default;
    }

    PerfUtils::PerfUtils()
        : _workthread(nullptr)
    {
        std::time_t now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
        tm utc_now;
        gmtime_s(&utc_now, &now);

        std::wstringstream oss;
        oss << L"cxx.calc.perflog-"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_mday << L"-"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_mon + 1 << L"-"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_year + 1900 << L"-"
            << L"_"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_hour << L"-"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_min << L"-"
            << std::setw(2) << std::setfill(L'0') << utc_now.tm_sec
            << L".log";

        auto filename = oss.str();
        auto folder = Windows::Storage::ApplicationData::Current->LocalFolder;
        create_task(
            folder->CreateFileAsync(ref new Platform::String(filename.c_str()), CreationCollisionOption::ReplaceExisting),
            task_continuation_context::use_arbitrary())
            .then([this](StorageFile ^ file) { _file = file; }, task_continuation_context::use_arbitrary())
            .wait();

        _workthread = new std::thread([this]() { _WorkThread(); });
    }

    void PerfUtils::WriteLine(const std::wstring& content)
    {
        WriteRequest req = { content };
        _SubmitReq(std::move(req));
        _cv.notify_all();
    }

    PerfUtils::ScopedLog PerfUtils::CreateScopedLog(const std::wstring& content)
    {
        return ScopedLog(*this, content);
    }

    void PerfUtils::_WorkThread()
    {
        _quit_flag.test_and_set(std::memory_order::memory_order_release);
        while (_quit_flag.test_and_set(std::memory_order::memory_order_acquire))
        {
            for (;;)
            {
                auto req = _FetchReq();
                if (!req.has_value())
                    break;

                create_task(
                    FileIO::AppendTextAsync(
                        _file,
                        ref new Platform::String(req.value().content.c_str())),
                    task_continuation_context::use_arbitrary())
                    .wait();
            }
            std::unique_lock<std::mutex> lk(_cv_mtx);
            _cv.wait(lk);
        }
    }

    void PerfUtils::_SubmitReq(WriteRequest&& req)
    {
        std::scoped_lock<std::mutex> lock(_wreqs_mtx);
        _wreqs.emplace_back(std::move(req));
    }

    std::optional<PerfUtils::WriteRequest> PerfUtils::_FetchReq()
    {
        std::scoped_lock<std::mutex> lock(_wreqs_mtx);
        if (_wreqs.size() > 0)
        {
            auto retval = std::make_optional<WriteRequest>(std::move(_wreqs.front()));
            _wreqs.pop_front();
            return retval;
        }
        return std::optional<WriteRequest>();
    }

    std::wstring PerfUtils::_TimeStamp()
    {
    }

    PerfUtils::ScopedLog::ScopedLog(PerfUtils& utils, const std::wstring& content)
        : _utils(&utils)
        , _message(content)
    {
        _utils->WriteLine(L"ScopedLog | " + _message + L" | begins");
    }

    PerfUtils::ScopedLog::~ScopedLog()
    {
        _utils->WriteLine(L"ScopedLog | " + _message + L" | ends");
    }

    PerfUtils::ScopedLog::ScopedLog(ScopedLog&& rhs)
        : _utils(std::move(rhs._utils))
        , _message(std::move(rhs._message))
    {}

    PerfUtils::ScopedLog& PerfUtils::ScopedLog::operator=(ScopedLog&& rhs)
    {
        _utils = std::move(rhs._utils);
        _message = std::move(rhs._message);
        return *this;
    }
}

